refactor(allocator/vec2): move `len` field from `Vec` to `RawVec`

Dunqing opened the pull request:
Prepare for #9706.

Move `len` field from `Vec` to `RawVec` to adapt upcoming change that reduces `len` and `cap` fields from `usize` to `u32` so that the `Vec` size can reduce from `32` to `24` without padding.

Marked `Vec` as `#[repr(transparent)]` and `RawVec` as `#[repr(C)]` to easier to solve raw transfer test failures because the fields are no longer reordered, we can statically determine the offsets, as mentioned in that comment.

Dunqing commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/10883?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#10883** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/10883?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/10883?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#10879** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/10879?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=10883).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

overlookmotel commented:
On second thoughts, maybe just moving `len` field into `RawVec` is the simpler option. If that field is `pub(super)`, it can still be accessed from `Vec`, so little else changes.

Boshen commented:
We should add a `assert_eq!(std::mem::size_of::<T>(), 32);` to these types.

Dunqing commented on crates/oxc_allocator/src/vec2/mod.rs:
```diff
@@ -955,14 +955,14 @@ impl<'bump, T: 'bump> Vec<'bump, T> {
+        let current_len = self.buf.len;
         unsafe {
-            let mut ptr = self.as_mut_ptr().add(self.len);
+            let mut ptr = self.as_mut_ptr().add(current_len);
```
Wrong!

Dunqing commented:
> We should add a `assert_eq!(std::mem::size_of::<T>(), 32);` to these types.

All of `generated/assert_layout.rs` is already included this assert actually.

overlookmotel commented:
There should be no perf impact from moving `len` field into `RawVec` in itself. I assume the perf difference is that field order changes in the process.

But to test that assumption (and make sure we're not getting false results from CodSpeed on either/both of these PRs), I suggest re-ordering the fields of `RawVec` in this PR as:

```rs
#[repr(C)]
pub struct RawVec<'a, T> {
    ptr: NonNull<T>,
    a: &'a Bump,
    cap: usize,
    pub(super) len: usize,
}
```

This is the same order the fields were in before this PR (with default `#[repr(Rust)]`, compiler moves fields with niches e.g. `&T` to be before fields without niches e.g. `usize`). I checked that's correct by adding this code (on main):

```rs
const _: () = {
    assert!(std::mem::size_of::<Vec<u8>>() == 32);
    assert!(std::mem::align_of::<Vec<u8>>() == 8);
    assert!(std::mem::offset_of!(Vec<u8>, buf) == 0);
    assert!(std::mem::offset_of!(Vec<u8>, len) == 24);
};

const _: () = {
    assert!(std::mem::size_of::<RawVec<u8>>() == 24);
    assert!(std::mem::align_of::<RawVec<u8>>() == 8);
    assert!(std::mem::offset_of!(RawVec<u8>, ptr) == 0);
    assert!(std::mem::offset_of!(RawVec<u8>, a) == 8);
    assert!(std::mem::offset_of!(RawVec<u8>, cap) == 16);
};
```

Those assertions all pass on main branch. If you re-arrange `RawVec`'s fields as suggested above, they should still pass on this PR.

After that change, this PR should have zero perf impact. If it still does, something is wrong.

overlookmotel commented on tasks/ast_tools/src/generators/raw_transfer.rs:
```diff
@@ -25,10 +25,9 @@ use crate::{
+// The `Vec` has marked `#[repr(C)]` layout, so the offsets are fixed.
 const VEC_PTR_FIELD_OFFSET: usize = 0;
-const VEC_LEN_FIELD_OFFSET: usize = 24;
+const VEC_LEN_FIELD_OFFSET: usize = 8;
```
This needs to go back to 24. Test NAPI CI task has not failed but it's still running. Probably it's in an infinite loop.

overlookmotel replied:
I've fixed this.

overlookmotel commented:
CI got stuck (as usual!). I've rebased on main, fixed the raw transfer test failure, and force pushed (ditto for #10884).

overlookmotel commented on tasks/ast_tools/src/generators/raw_transfer.rs:
```diff
@@ -25,8 +25,7 @@ use crate::{
-// They will be fully reliable once we implement our own `Vec` type and make it `#[repr(C)]`.
+// `Vec` is `#[repr(transparent)]` and `RawVec` is `#[repr(C)]`, so these offsets are fixed.
 const VEC_PTR_FIELD_OFFSET: usize = 0;
 const VEC_LEN_FIELD_OFFSET: usize = 24;
```
Completely, and bizarrely, wrong! `len` is after `cap` field, so at offset 24.

overlookmotel commented:
### Merge activity

* **May 15, 5:48 PM EDT**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

Dunqing committed:
refactor(allocator/vec2): move `len` field from `Vec` to `RawVec` (#10883)

Prepare for #9706.

Move `len` field from `Vec` to `RawVec` to adapt upcoming change that reduces `len` and `cap` fields from `usize` to `u32` so that the `Vec` size can reduce from `32` to `24` without padding.

Marked `Vec` as `#[repr(transparent)]` and `RawVec` as `#[repr(C)]` to easier to solve raw transfer test failures because the fields are no longer reordered, we can statically determine the offsets, as mentioned in that comment.
